gossipsub.go

```diff
@@ -70,6 +77,62 @@ func NewGossipSub(ctx context.Context, h host.Host, opts ...Option) (*PubSub, er
 	return NewPubSub(ctx, h, rt, opts...)
 }
 
+// WithPeerScore is a gossipsub router option that enables peer scoring.
+//
+// gossipThreshold is the score threshold below which gossip propagation is supressed.
+//
+// publishThreshold is the score threshold below which we shouldn't publish when using flood
+//  publishing (also applies to fanout and floodsub peers).
+//
+// graylistThreshold is the score threshold below which message processing is supressed altogether,
+//  implementing an effective graylist according to peer score.
+//
+// These thresholds should generally be negative, allowing some information to disseminate from low
+//  scoring peers.
```

Comment:
Let's add checks for these and emit a warning please, e.g., it seems that almost always we would want a graylist threshold below 0 and any value above this range will likely be a programmer error.

Reply:
From reading the spec it seems we should also check that gossip > publish > graylist (especially because this relation doesn't evident from reading the code).

Reply:
Yeah, I'll add a check in validate.

Reply:
done.